qualtran/bloqs/arithmetic/addition.py

```diff
@@ -152,63 +171,81 @@ def _right_building_block(self, inp, out, anc, depth):
         else:
             yield CNOT().on(anc[depth - 1], anc[depth])
             yield And().adjoint().on(inp[depth], out[depth], anc[depth])
-            yield CNOT().on(anc[depth - 1], inp[depth])
-            yield CNOT().on(inp[depth], out[depth])
+            # In the _left_building_block we sometimes use an
+            # ancilla in place of the input bit. We do this when the
+            # input bit is None, therefore no uncomputation is necessary.
+            if anc[depth - 1] != inp[depth]:
+                yield CNOT().on(anc[depth - 1], inp[depth])
+            else:
+                inp[depth] = None
+            if inp[depth] is not None:
+                yield CNOT().on(inp[depth], out[depth])
             yield from self._right_building_block(inp, out, anc, depth - 1)
 
     def decompose_from_registers(
         self, *, context: cirq.DecompositionContext, **quregs: NDArray[cirq.Qid]
     ) -> cirq.OP_TREE:
         # reverse the order of qubits for big endian-ness.
         input_bits = quregs['a'][::-1]
+        # Make len(input_bits) = len(output_bits) by appending Nones
+        input_bits = np.append(input_bits, [None] * (self.b_dtype.bitsize - self.a_dtype.bitsize))
```

Comment:
since the assumption is that `self.b_dtype.bitsize >= self.a_dtype.bitsize`, please add a validator for that